### PR TITLE
Apply new terminology to the merge spec

### PR DESCRIPTION
**Pull request by mkalinin**
Updates the spec according to the new terminology. 

There is a decision to rename `application-layer` to `execution-layer` to avoid potential confusion with the layer where smart contracts and apps built around them are living.

There is an essay utilising new terms to give an idea of how they should be used
- https://hackmd.io/@n0ble/the-merge-terminology

**UPD**

In addition, it renames `ExecutionBlockHeader` to `ExecutionPayloadHeader` to emphasize the difference between the payload header and pre-merge PoW `BlockHeader`.

### CODE REVIEW

File: specs/merge/beacon-chain.md [hunk 1]
```diff
@@ -138,14 +138,14 @@ class ApplicationBlockHeader(Container):
...
-    return state.latest_application_block_header.block_hash != Bytes32()
+    return state.latest_execution_payload_header.block_hash != Bytes32()
 ```
 
 #### `is_transition_block`
 
 ```python
 def is_transition_block(state: BeaconState, block_body: BeaconBlockBody) -> boolean:
-    return state.latest_application_block_header.block_hash == Bytes32() and block_body.application_payload.block_hash != Bytes32()
+    return state.latest_execution_payload_header.block_hash == Bytes32() and block_body.execution_payload.block_hash != Bytes32()
```
Comment (hwwhww) [inline specs/merge/beacon-chain.md:148]: The linter max-line length is 120 now.
Suggesting:
```suggestion
    is_empty_latest_execution_payload_header_block_hash = state.latest_execution_payload_header.block_hash == Bytes32()
    is_empty_execution_payload_block_hash = block_body.execution_payload.block_hash == Bytes32()
    return is_empty_latest_execution_payload_header_block_hash and not is_empty_execution_payload_block_hash
```

File: specs/merge/beacon-chain.md [hunk 2]
```diff
@@ -156,54 +156,54 @@ def process_block(state: BeaconState, block: BeaconBlock) -> None:
...
 
-Let `get_application_state(application_state_root: Bytes32) -> ApplicationState`  be the function that given the root hash returns a copy of ethereum application state. 
+Let `get_execution_state(execution_state_root: Bytes32) -> ExecutionState`  be the function that given the root hash returns a copy of ethereum execution state. 
 The body of the function is implementation dependent.
 
-##### `application_state_transition`
+##### `execution_state_transition`
 
-Let `application_state_transition(application_state: ApplicationState, application_payload: ApplicationPayload) -> None` be the transition function of ethereum application state. 
+Let `execution_state_transition(execution_state: ExecutionState, execution_payload: ExecutionPayload) -> None` be the transition function of ethereum execution state.
```
Comment (hwwhww) [inline specs/merge/beacon-chain.md:173]:
```suggestion
Let `execution_state_transition(execution_state: ExecutionState, execution_payload: ExecutionPayload) -> None` be the transition function of Ethereum execution state. 
```

**Review by hwwhww — APPROVED** [review]
The rebased #2316 is here: https://github.com/ethereum/eth2.0-specs/commit/24f6521751869590d8975793767cafc0a20b0fbc

Added comments on the proposed spec changes here. I'll update #2316 once this PR is merged. Otherwise, LGTM! 👍

**Comment by mkalinin** [comment]
@hwwhww I have addressed all the comments you've made, thanks! Also, fixed a bug in checking pre-conditions of `process_execution_payload` function.

File: specs/merge/beacon-chain.md [hunk 3]
```diff
@@ -156,54 +156,52 @@ def process_block(state: BeaconState, block: BeaconBlock) -> None:
...
-        receipt_root=application_payload.receipt_root,
-        logs_bloom=application_payload.logs_bloom,
-        transactions_root=hash_tree_root(application_payload.transactions),
+    execution_payload = body.execution_payload
+
+    if is_transition_completed(state):
+        assert execution_payload.parent_hash == state.latest_execution_payload_header.block_hash
+        assert execution_payload.number == state.latest_execution_payload_header.number + 1
+
+    execution_state = get_execution_state(state.latest_execution_payload_header.state_root)
```
Comment (djrtwo) [inline specs/merge/beacon-chain.md:192]: Don't we need to skip this execution state transition if we are before the merge transition?

Maybe a blocking condition at the top?

**Review by djrtwo — COMMENTED** [review]
looks good! one question about `process_execution_payload` before the transition

File: specs/merge/beacon-chain.md (hunk 3 above)
Comment (mkalinin) [reply specs/merge/beacon-chain.md:192]: Definitely, I have completely missed this code branch. Let me fix

**Comment by mkalinin** [comment]
> looks good! one question about `process_execution_payload` before the transition

Fixed now 👍

**Review by djrtwo — APPROVED** [review]

**Pull request merged**